// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto2";

package cc.proto;

option optimize_for = LITE_RUNTIME;

// Identifies the type of cc:Layer a LayerNode represents. It is used to
// facilitate reconstruction of a Layer of the correct type on the client.
enum LayerType {
  Base = 1;

  // TODO(nyquist): Add the rest of the necessary LayerTypes.
};

// Hierarchical structure for serializing the Layer tree.
message LayerNode {
  // required
  optional int32 id = 1;
  // required
  optional LayerType type = 2;
  optional int32 parent_id = 3;
  // A List of all the children of the current LayerNode.
  repeated LayerNode children = 4;
  optional LayerNode mask_layer = 5;
  optional LayerNode replica_layer = 6;
}

// A container for a list of dirty layers.
message LayerUpdate {
  // A list of dirty layers.
  repeated LayerProperties layers = 1;
}

message LayerProperties {
  // required
  optional int32 id = 1;
  // required
  optional bool needs_push_properties = 3;
  // required
  optional int32 num_dependents_need_push_properties = 4;

  // The properties below are only read if |needs_push_properties| is set.
  // The Layer base class and each descendant have different proto messages
  // for their specific properties.
  optional BaseLayerProperties base = 5;
}

message BaseLayerProperties {
  // TODO(nyquist): Add all the required properties below. Huzzah!
}
